doc/developer/platform/architecture-storage.md

```diff
@@ -107,25 +107,150 @@ All commands are durably persisted and their effect survives crashes and reboots
+    It handles contended writes, picking at most one winner, and providing a single consistent history of updates for each shard.
+    A collection may be formed of multiple shards, which may not have the same durable frontiers.
+
+3. We have an evolving understanding of what should happen to our data from raw sources to `persist`.
+
+    We want some amount of flexibility here, as we may be surprised by the requirements of source decoding, or new opportunities.
+    The "understanding" should reflect deterministic computation.
+    Non-determinism is something we can do with an advanced recipe (more `persist` stuff, in the spirit of reclocking, where we record non-deterministic choices).
+
+### The recipe
+
+There is some prep work to do first:
+
+1.  Identify all "raw sources" that contribute to some intended output (e.g. kafka topics, metadata, schema registry).
+
+2.  "Reclock" each input, by choosing and recording in `persist` shards a collection of source addresses that are monotonically increasing (do not go backwards, but needn't always go forward).
+
+    We should use times that roughly track "real time" so that they align with the times of other collections.
+    The reclocked inputs are going to be treated as if their data "become available" at the first moment for which the data's address is covered by the reclocking collection.
+
+    As suggested, this may include things like the schema registry that aren't conventional inputs, but nonetheless are external sources that influence the output we will produce.
+
+2. Identify the "state" that a streaming implementation of the intended functionality would need to maintain.
+
+    For example, if we are matching up transaction metadata with the data it references, we may need to maintain some outstanding metadata records and some outstanding data records, until all in a transaction match.
+    If we intend to hold back transactions until all strictly prior transactions have completed, we may need to hold on even to some completed transactions, awaiting the completion of the prior work.
+
+3. Create `persist`-backed storage collections for the output and the identified "state".
+
+
+With this prep work done, the assembly instructions are:
+
+1. To start, determine `t` the `min` write frontier in `persist` for all shards of outputs and all state.
+
+    The time `t` may be strictly less than some of the shards, which will mean that we cannot necessarily write at `t` for all shards.
+    This is fine, and will be discussed in just a moment.
+
+2. From time `t` onward, rehydrate the "state" associated with the source pipeline.
+
+3. Restart each raw source from the addresses found in the persisted reclocking collection at `t`.
+
+4. From time `t` onward, continually produce reclocked messages: messages from the raw source whose addresses are translated to timestamps using reclocking information from `persist`.
```

Comment:
hmm do you think something like the current debezium transaction implementation will cause issues here: currently the logic timestamp comes from the metadata information -- the timestamp assigned to each data message is essentially ignored.  If we do the same thing here, we're just completely ignoring the persisted reclocking collection

Reply:
My hunch is that this will fight with the current DBZ approach. We might need to put all the constraints on the table, and now might be a good time to pre-think that before concluding that the pipeline is generally valid. For example, if we want to reclock data sources and then use their times to (in part) constrain the release of complete transactions, this makes it harder to emit completed transactions in order (each source would need to be aware of incomplete transactions across all tables, not just their own).

Repurposing the current DBZ pipeline would mean (I think) putting the consistency topic in charge of the reclocking of the data inputs (it would determine when certain data "are visible"), and it .. might do weird things like announce data as available before it is actually present in the topic. Or, .. it would need to somehow play the data out while looking at the transaction timestamps in order to determine the reclocking association.

Another take could be that this would require a reclocking of the reclocked data, into a timeline associated with the transactions.

The reasons we reclock, btw, aren't just progress reporting. It's there so that we can resume on a consistent cut through the inputs, and know how much or little of each input we need to retain / where to start from on a reboot. If we don't have this written down some way, it's a lot more custom to figure out how to recover correctly, and from where, and what of the input needs to be recorded to make this happen.

Reply:
> Another take could be that this would require a reclocking of the reclocked data, into a timeline associated with the transactions.

Doing this conceptually checks out to me -- especially given your reminder about needing the reclocking for resumption purposes.  That the transaction time domain corresponds to the metadata topic time domain would be incidental but not strictly necessary for correctness of progress / resumption